### ISSUE TITLE
3D Secure 2.0 Support

### ISSUE THREAD

**Issue by BassemN** [OpenMage/magento-lts#812]
Hello,

Based on this [article](https://community.magento.com/t5/Magento-DevBlog/3D-Secure-2-0-changes/ba-p/136460)
> 3DS 1.0 is supported. When and/or where use of 3DS 2.0 is required, Merchants will either need to replace PayPal with Braintree or upgrade to Magento 2.3.x.

Magento 1 will not support 3DS 2 for PayPal pro so will this repo support the 3DS 2?

Thanks

**Comment by rudracomputech** [comment]
Hello Team, 
Do we have a chance to get 3ds with openMage if yes they when ?
Thanks

**Comment by justinbeaty** [comment]
@rudracomputech I would try to use the Gene Braintree module.

If you use composer, I have a fork of the module with the latest Braintree SDK here: https://github.com/justinbeaty/module-gene-braintree

If you do not use composer, you could download the old version here: https://github.com/justinbeaty/module-gene-braintree/archive/e0793fb2861abde2084d285cd8bb7653ad3dd640.zip

I am not sure if the module's built-in javascript files supports 3DS v2. I do know that the PHP sdk at least supports it, because I have 3DS v2 working but using some custom javascript.

**Comment by rudracomputech** [comment]
Hi @justinbeaty  thankyou for your response. I tried with your module but no luck I am getting "Card payment has failed, missing token/nonce" in log. 
I am using 
<pre>

Magento Braintree Payments Version	3.1.3
You're up to date!
You're running the latest version of our module.

For questions or assistance with the module, please visit our Support Portal.
Braintree PHP SDK Version	3.23.0
</pre>

Thanks in advance

**Comment by justinbeaty** [comment]
@rudracomputech Okay so it looks like you are using the ZIP file (aka the last official release from Gene).

For your issue, maybe it could be incorrect configuration in the Magento Admin Panel? Look under Sales -> Payment Methods.

First, does it say "Valid Credentials - You're ready to accept payments via Braintree" ? Second, is the "Braintree Payments - Credit Card" method enabled?

Finally, the last thing I ask you to check is if there are any JS error on the frontend during checkout. It is possible that the JS api Gene originally used no longer works. I would not know for certain because I am using a custom frontend where I import Braintree's latest web framework. In other words I do not use Magento's default checkout flow.

**Comment by rudracomputech** [comment]
@justinbeaty  yes it says valid credentials in admin and also credit card is enable. what I found the field "payment_method_nonce" passing empty value
please check https://beta.parkerbrand.co.uk/onepage/

**Comment by justinbeaty** [comment]
@rudracomputech I did try and checkout and I get the error. I think this is maybe a problem with your one page checkout customizations. Are you using a 3rd party checkout module, or just customized version of default Magento's checkout? Can you temporarily switch to the default onepagecheckout module from stock OpenMage just to debug? 

To diagnose, I checked Chrome's network log and there is no call to Braintree's API. Normally what should happen is that on click of the "continue" button, the Braintree module should call Braintree's API to generate a nonce for the payment, but this does not seem to be happening.

**Comment by fballiano** [comment]
I'm using gene_braintree on a customer of mine, with firecheckout, and it's still working fine with creditcards+paypal, does this help you in any way or could you need more info?

**Comment by justinbeaty** [comment]
@fballiano It does look like Gene has support for firecheckout:

```
user@local:~/modules/module-gene-braintree-public$ grep -Ri firecheckout app/
app/design/frontend/base/default/template/gene/braintree/js/firecheckout.phtml:<!-- FIRECHECKOUT BRAINTREE SUPPORT -->
```

But, @rudracomputech is this the checkout module you're using?

Also, if you would like @fballiano this can maybe be moved to discussions instead of an issue. (I think you may be able to move it with your repo permissions, not sure).
